scripts/igmp_query_injector.py

```diff
@@ -0,0 +1,459 @@
+        pass
+
+
+class VifsInjector(IGMPQueryInjector):
+    def __init__(self, max_resp_time, vifs, vif_connected_timeout=0):
+        super(VifsInjector, self).__init__(max_resp_time)
+        self.vifs = vifs
+        self.vif_connected_timeout = vif_connected_timeout
+
+    def inject(self):
+        if self.vif_connected_timeout > 0:
+            # should check connection state
+            log.info('Inject IGMP query with connection state check')
+            self._inject_with_connection_state_check()
+        else:
+            log.info('Inject IGMP query without connection state check')
+            self._inject_without_connection_state_check()
+
+    def _inject_with_connection_state_check(self):
+        checkers = self._build_vif_state_checkers()
+        with self._start_watcher(checkers) as (watcher, poll):
+            remain_timeout = self.vif_connected_timeout
+            while remain_timeout > 0:
+                start = time.time()
+                events = poll.poll(remain_timeout)
+                if not events:
+                    break
+
+                for _ in events:
+                    path = watcher.readline()
+                    if path not in checkers:
+                        continue
+
+                    checker = checkers[path]
+                    if checker.state_is_expected():
+                        self.inject_to_vif(checker.vif, checker.mac)
+                        del checkers[path]
+
+                if not checkers:
+                    # all of the vif state has changed to connected
+                    break
+
+                remain_timeout -= (time.time() - start)
+
+        for checker in checkers.itervalues():
+            log.warning("Value of '%s' not change to '%s' in %d seconds." %
+                        (checker.vif_state_path, checker.expect_state, self.vif_connected_timeout))
+            log.warning("Won't inject IGMP query to vif: %s, mac: %s" % (checker.vif, checker.mac))
+
+    def _build_vif_state_checkers(self):
+        vif_state_connected = '4'
+        checkers = [VifStateChecker(vif, vif_state_connected) for vif in self.vifs]
+        return dict([(checker.vif_state_path, checker) for checker in checkers])
+
+    @contextmanager
+    def _start_watcher(self, checkers):
+        watcher = XenstoreWatcher(checkers.keys())
+        watcher.start()
+        poll = epoll()
+        poll.register(watcher.stdout, EPOLLIN)
+        yield watcher, poll
+
+        poll.close()
+        watcher.terminate()
+
+    def _inject_without_connection_state_check(self):
+        self.inject_to_vifs(self.vifs)
+
+
+class BridgeInjector(IGMPQueryInjector):
+    RE_VIF = re.compile(r'^vif\d+\.\d+$')
+
+    def __init__(self, max_resp_time, bridges):
+        super(BridgeInjector, self).__init__(max_resp_time)
+        self.bridges = bridges
+
+    def get_vifs_on_bridge(self, bridge):
+        ret = []
+        outs = subprocess.check_output(['ovs-vsctl', 'list-ports', bridge]).strip()
```

Comment:
It is not ideal to find the list of vifs by searching the output of `ovs-vsctl`.   We should try to find a neater way of doing this from other datasources such as sysfs, to avoid problems if the names of vif interfaces change in the future.

Reply:
I think the name of the **vif** devices should fixed starts with `vif`.
Refer to: https://github.com/xapi-project/xenopsd/blob/master/xc/xenops_server_xen.ml#L2563

Reply:
This is true at the moment, and we do rely on this in some other places, so I would not mind this too much. However, it would be better not to if we can avoid it. Another option could be to allow for multiple VIF arguments on the command line, rather than the `bridge` option, and let networkd determine which VIFs are on the bridge (network already has helper functions for this).

Reply:
@robhoes My concern is if the bridge has too much vifs and we pass the vifs as arguments, will the length of the command exceeds the limitation of  shell command line?